### DISCUSSION TITLE
inference batchsize

### DISCUSSION THREAD

**Discussion by WeissShahaf** [talmolab/sleap#1574] [Ideas]
Dear Devs,

I'm trying to do inference on a very long movie (days), and i'm while the model is good. i'm getting ~40FPS during inference.
my GPU is only using 30% cuda and 2GB of VRAM. 
where in the code can i change the batch size for infrence?
maybe make that a variable in the GUI?

thanks,
Shahaf

**Comment by roomrys** [comment]
So there is a `--batch_size` argument for the [`sleap-track`](https://sleap.ai/guides/cli.html#sleap-track) CLI which means you can try running inference via the CLI while messing with batch size to see if that helps much. As for implementing it in the GUI...

## Where is batch size used?

1. All of the `Predictor`s are sub-classed off the abstract `Predictor` base class. Each `Predictor` is required to implement a `from_trained_models` method which takes as input a `batch_size` argument.
2. The type of `Predictor` to create is delegated by the `Predictor.from_model_paths` method which calls the correct `from_trained_models` for that model type.
3. When running inference (either through the GUI or CLI), the predictor is created via the `load_model` function
4. which is called by `_make_predictor_from_cli`
5. which is called by the `main` function in sleap.nn.inference (which is the entry point for the `sleap-track` command).

## How does this interact with the GUI

1. The GUI uses a subprocess in the `InferenceTask.predict_subprocess` method to call a CLI command (`sleap-track`),
2. after gathering all the CLI arguments from the user in the `InferenceTask.make_predict_cli_call` method.
3. (Most of) the CLI args are stored in the `item_for_instance: ItemsForInferece` variable which is passed into both `make_predict_cli_call` and `predict_subprocess`
4. So where does `item_for_inference` come from? Well from `items_for_inference.items`, duh.
5. So where does `items_for_inference` come from? We see it is passed into `run_gui_inference` (which calls `predict_subprocess` on each item).
6. `items_for_inference` is also passed into `run_learning_pipeline` (which calls `run_gui_inference`).
7. Aha, `items_for_inference` is gathered when running `LearningDialog.get_items_for_inference` taking as a single input the `pipeline_form_data`.
8. The `pipeline_form_data` is obtained just a line above using the `LearningDialog.pipeline_from_widget.get_form_data` method (which just gets all the data from the GUI created via the `pipeline_form.yaml`).

## What should be edited?

So, to finally answer the question of what to change... 

- [ ] We need to add a to-be-GUI-element to the `pipeline_form.yaml` under `inference` for the batch size, maybe something similar to the batch size for training?
https://github.com/talmolab/sleap/blob/dbe14a8f860f2aecfdd34bce802cde03db6be000/sleap/config/training_editor_form.yaml#L654-L664

- [ ] We also need to modify some code in the `LearningDialog.get_items_for_inference` to get the `batch_size` from the `pipeline_form_data`
https://github.com/talmolab/sleap/blob/dbe14a8f860f2aecfdd34bce802cde03db6be000/sleap/gui/learning/dialog.py#L580-L584

- [ ] Modify code in `LearningDialog.get_items_for_inference` to use `batch_size` to create `items` for inference in all options for `items_for_inference`  which seems like we would also need to:
- [ ] Modify `DatasetItemForInference`
- [ ] Modify `ItemsForInference.from_video_frames_dict`
https://github.com/talmolab/sleap/blob/dbe14a8f860f2aecfdd34bce802cde03db6be000/sleap/gui/learning/dialog.py#L586-L611
